File: 2020/py/day2.py

def main():
	f = open("/home/detmer/programming/aoc/2020/aoc_input/input_day2.txt", "r")
	valid_pass_part1 = 0
	valid_pass_part2 = 0

	for line in f:

		#splitting the strings into workable parts

		pass_split = line.split(":")
		pol_split = pass_split[0].split(" ")
		polnum_split = pol_split[0].split("-")
		
		#naming vars for readability

		minimum = int(polnum_split[0])
		maximum = int(polnum_split[1])
		letter = pol_split[1]
		password = pass_split[1]

		#count occurences of letter in password

		amount = 0

		for i in password:
			if i == letter:
				amount+=1

		#check if password is valid according to the password policy

		if(amount >= minimum and amount <= maximum):
			valid_pass_part1+=1

		#part two
		#checking if the character is at specific location according to the password policy
		if password[minimum] == letter and password[maximum] != letter or password[minimum] != letter and password[maximum] == letter:
			valid_pass_part2+=1

	print("Valid passwords part one: " + str(valid_pass_part1))
	print("Valid passwords part two: " + str(valid_pass_part2))


if __name__ == '__main__':
	main()

File: 2020/c/day2.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

int main(void){

	FILE* filePointer;
	int bufferLength = 255;
	char buffer[bufferLength];
	char* pol;
	char* minimum;
	char* maximum;
	char* rest;
	char* letter;
	char* password;
	int valid_pass = 0;
	int valid_pass_two = 0;

	filePointer = fopen("/home/detmer/programming/aoc/2020/aoc_input/input_day2.txt", "r");

	while(fgets(buffer, bufferLength, filePointer)) {
		pol = strtok(buffer, ":");
		password = strtok(NULL, ":");
		minimum = strtok(pol,"-");
		rest = strtok(NULL,"-");
		maximum = strtok(rest, " ");
		letter = strtok(NULL," ");
		int count = 0;
		for(int i=0; i < strlen(password); i++){
			if(password[i] == letter[0]){
				count++;
			}
		}

		int min = atoi(minimum);
		int max = atoi(maximum);

		if(count >= min && count <= max){
			valid_pass++;
		}

		//part two
		if(password[min] != letter[0] && password[max] == letter[0] || password[min] == letter[0] && password[max] != letter[0]){
			valid_pass_two++;
		}

	}
	printf("The amount of valid passwords is part one : %d\n", valid_pass );
	printf("The amount of valid passwords is part two : %d\n", valid_pass_two );

	fclose(filePointer);
}